upgrades/backend.go

```diff
@@ -42,3 +55,129 @@ func (s stateBackend) AddVirtualHostKeys() error {
 func (s stateBackend) SplitMigrationStatusMessages() error {
 	return state.SplitMigrationStatusMessages(s.pool)
 }
+
+func (s stateBackend) PopulateApplicationStorageUniqueID() error {
+	return state.PopulateApplicationStorageUniqueID(s.pool, GetStorageUniqueID(NewK8sClient))
+}
+
+func NewK8sClient(model *state.Model) (kubernetes.Interface, *rest.Config, error) {
+	g := stateenvirons.EnvironConfigGetter{Model: model}
+	cloudSpec, err := g.CloudSpec()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	cfg, err := jujukubernetes.CloudSpecToK8sRestConfig(cloudSpec)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	k8sClient, err := kubernetes.NewForConfig(cfg)
+	return k8sClient, cfg, err
+}
+
+type newK8sFunc func(model *state.Model) (kubernetes.Interface, *rest.Config, error)
+
+type kubernetesClient struct {
+	kubernetes.Interface
+	config *rest.Config
+}
+
+func GetStorageUniqueID(newK8sClient newK8sFunc) func(appName string, model *state.Model) (string, error) {
+	k8sClientsByModel := make(map[string]kubernetesClient)
+	return func(appName string, model *state.Model) (string, error) {
```

Comment:
Let's pass in a slice of app names for the model and return a slice of storage ids.
That way we only create the k8s client once per model.